text/0000-structural-records.md

```diff
@@ -0,0 +1,1262 @@
+  }
+  ```
+
++ For `PartialEq`, each field is compared with same field in `other: Self`.
+
++ For `ParialOrd` and `Ord`, lexicographic ordering is used based on
+  the name of the fields and not the order given because structural records
+  don't respect the order in which the fields are put when constructing or
+  giving the type of the record.
+
++ For `Debug` the same lexicographic ordering for `Ord` is used.
+  As an example, when printing out `my_record` as with:
+
+  ```rust
+  let my_record = { beta: true, alpha: 42u8, gamma: "Dancing Ferris" };
+  println!("{:#?}", my_record);
+  ```
+  
+  the following would appear:
+  
+  ```rust
+  {
+      alpha: 42,
+      beta: true,
+      gamma: "Dancing Ferris"
+  }
+  ```
+
++ For `Hash`, the same ordering of the fields as before is used and then
+  `self.the_field.hash(state)` is called on each field in that order.
+  For example:
+
+  ```rust
+  self.alpha.hash(state);
+  self.beta.hash(state);
+  self.gamma.hash(state);
+  ```
+
+For auto traits (e.g. `Send`, `Sync`, `Unpin`), if all field types implement
+the auto trait, then the structural record does as well.
+For example, if `A` and `B` are `Send`, then so is `{ x: A, y: B }`.
+
+A structural record is `Sized` if all field types are.
+If the lexicographically last field of a structural record is `!Sized`,
+then so is the structural record. If any other field but the last is
+`!Sized`, then the type of the structural record is not well-formed.
+
+### Implementations and orphans
+
+[RFC 2451]: https://github.com/rust-lang/rfcs/pull/2451
+
+It is possible to define your own implementations for a structural record.
+The orphan rules that apply here are those of [RFC 2451] by viewing a structural
+record as a positional tuple after sorting the elements lexicographically.
+
+For example, if a `trait` is crate-local, we may implement it for a record:
+
+```rust
+trait Foo {}
+
+impl Foo for { alpha: bool, beta: u8 } {}
+```
+
+Under 2451, we can also write:
+
+```rust
+struct Local<T>(T);
+
+impl From<()> for Local<{ alpha: bool, beta: u8 }> { ... }
+```
+
+This is valid because `Local` is considered a local type.
+
+However, a structural record itself isn't a local type, so you cannot write:
+
+```rust
+struct A;
+
+impl From<()> for { alpha: A, beta: u8 } { ... }
+```
+
+This is the case even though `A` is a type local to the crate.
+
+The behaviour for inherent implementations is also akin to tuples.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Grammar
+[grammar]: #grammar
+
+Given:
+
+```rust
+field ::= ident | LIT_INTEGER ;
+```
+
+We extend the expression grammar with:
+
+```rust
+expr ::= ... | expr_srec ;
+
+expr_srec ::= "{" field_init "," field_inits_tail? "}" ;
+field_inits_tail ::= field_inits ","? | (field_inits ",")? DOTDOT expr ;
+field_inits ::= field_init ("," field_init)* ;
+field_init ::= ident | field ":" expr ;
+```
+
+Note that this grammar permits `{ 0: x, 1: y }`.
```

Comment:
https://github.com/rust-lang/rfcs/issues/1595 was closed, so I'd expect this to be rejected.
This saves us from stuff like "checking that there are no gaps in the positional fields" as well.

Reply:
Yeah probably; I've left an unresolved question about it.

Reply:
@varkor yeah that was accepted to make macro writing easier.

Reply:
@petrochenkov I am not sure I understand the reasoning why this is not allowed:

```rust
struct Foo {
    0: i32, // This is the only thing that fails ...
    arg_1: i32,
    named_1: i32,
}

fn main() {
    let foo = Foo { 0: 0, arg_1: 1, named_1: 42} ;
    println!("{}", foo.0);
}
```

fails only at the struct definition with "^ expected identifier". Would it not just be necessary to change 'IDENTIFIER' to 'IDENTIFIER_OR_NUMBER' (pseudo-code wise)?

to make it all work as everything else seems to be able to work with '.0', '0: ', etc.?

I am also not seeing the drawback given that it is already inconsistent ...

My personal motivation with that is to mirror named function arguments in structs as they then perfectly match the: 

`0: arg_0, 1: arg_1, named_1: arg_3`

pattern. Also getting that in independent of structural records is one less thing that structural records need to handle.